### ISSUE TITLE
Partial mode should support syntactic diagnostics

### ISSUE THREAD

**Issue by amcasey** [microsoft/TypeScript#44835]
# Suggestion

## 🔍 Search Terms

partial mode, syntax-only, syntax diagnostics

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

When the server is in syntax-only mode, it should be able to provide syntactic diagnostics.  It would be non-trivial to provide all syntactic diagnostics because some are deferred until semantic checking, but we can and should report simple problems like unbalanced quotes and brackets.

For editors that consume diagnostics synchronously, this means supporting `syntacticDiagnosticsSync`.  For editors that consume diagnostics asynchronously this means accepting `getErr` and firing `syntaxDiag` events (but not `semanticDiag` or `suggestionDiag` events).

It's up to the [editor](https://github.com/microsoft/vscode/issues/127700) to decide when diagnostics are requested, but they probably should be limited to permanent partial mode (as opposed to temporary partial mode, during project load).

## 📃 Motivating Example

```ts
if ((A && B) || (!C && D) || (!E && F))) { 
   // There is an unbalanced paren above and it would be nice to find out without rebuilding
}
```

**Comment by sandersn** [comment]
@DanielRosenwasser  I don't understand the labels on this feature. 
1. This doesn't sound like an RC feature. Did we decide on that at the editor meeting?
2. If it's help wanted, do I need to work on it?

I haven't worked with partial mode at all, so it's going to take me more time than usual so I can learn how that works and how to test it.
